raid/xor_gen_avx.asm

```diff
@@ -99,43 +99,42 @@ default rel
 
 section .text
 
-align 16
+align 32
 mk_global  xor_gen_avx, function
 func(xor_gen_avx)
 
 	FUNC_SAVE
 	sub	vec, 2			;Keep as offset to last source
 	jng	return_fail		;Must have at least 2 sources
-	cmp	len, 0
+	test	len, len
 	je	return_pass
-	test	len, (128-1)		;Check alignment of length
+	test	BYTE(len), (128-1)	;Check alignment of length
 	jnz	len_not_aligned
 
 
 len_aligned_128bytes:
-	sub	len, 128
-	mov	pos, 0
+	add	len, -128
+	mov	DWORD(pos), DWORD(pos)
 
 loop128:
-	mov	tmp, vec		;Back to last vector
 	mov	tmp2, [arg2+vec*PS]	;Fetch last pointer in array
-	sub	tmp, 1			;Next vect
+	lea	tmp, [vec-1]		;Next vect
 	XLDR	ymm0, [tmp2+pos]	;Start with end of array in last vector
 	XLDR	ymm1, [tmp2+pos+32]	;Keep xor parity in xmm0-7
 	XLDR	ymm2, [tmp2+pos+(2*32)]
 	XLDR	ymm3, [tmp2+pos+(3*32)]
 
 next_vect:
 	mov 	ptr, [arg2+tmp*PS]
-	sub	tmp, 1
 	XLDR	ymm4, [ptr+pos]		;Get next vector (source)
 	XLDR	ymm5, [ptr+pos+32]
 	XLDR	ymm6, [ptr+pos+(2*32)]
 	XLDR	ymm7, [ptr+pos+(3*32)]
-	vxorpd	ymm0, ymm0, ymm4	;Add to xor parity
-	vxorpd	ymm1, ymm1, ymm5
-	vxorpd	ymm2, ymm2, ymm6
-	vxorpd	ymm3, ymm3, ymm7
+	vpxor	ymm0, ymm0, ymm4	;Add to xor parity
```

Comment:
This instruction can only be executed on systems that support AVX2, where this function implements AVX only instructions. vxorpd is an AVX instruction supporting ymms (https://www.felixcloutier.com/x86/xorpd)  , whereas vpxor only supports ymms if AVX2 is supported (https://www.felixcloutier.com/x86/pxor).